File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/interactors/CardsInteractorImplTest.java
package com.dbottillo.mtgsearchfree.interactors;

import com.dbottillo.mtgsearchfree.model.Deck;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.model.MTGSet;
import com.dbottillo.mtgsearchfree.model.SearchParams;
import com.dbottillo.mtgsearchfree.model.storage.CardsStorageImpl;
import com.dbottillo.mtgsearchfree.util.Logger;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import java.util.Arrays;
import java.util.List;

import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CardsInteractorImplTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    private static final int MULTIVERSE_ID = 180607;
    private CardsInteractor underTest;

    @Mock
    private CardsStorageImpl cardsStorageImpl;
    @Mock
    private MTGSet set;
    @Mock
    private SearchParams searchParams;
    @Mock
    private Deck deck;
    @Mock
    private MTGCard card;
    @Mock
    private MTGCard otherSideCard;
    @Mock
    Logger logger;

    private List<MTGCard> luckyCards = Arrays.asList(new MTGCard(2), new MTGCard(3));
    private List<MTGCard> favCards = Arrays.asList(new MTGCard(3), new MTGCard(4));
    private List<MTGCard> setCards = Arrays.asList(new MTGCard(5), new MTGCard(6));
    private List<MTGCard> searchCards = Arrays.asList(new MTGCard(6), new MTGCard(7));
    private List<MTGCard> deckCards = Arrays.asList(new MTGCard(7), new MTGCard(8));

    @Before
    public void setup() {
        when(cardsStorageImpl.getLuckyCards(2)).thenReturn(luckyCards);
        when(cardsStorageImpl.getFavourites()).thenReturn(favCards);
       // when(cardsStorageImpl.load(set)).thenReturn(setCards);
        when(cardsStorageImpl.doSearch(searchParams)).thenReturn(searchCards);
        when(cardsStorageImpl.loadDeck(deck)).thenReturn(deckCards);
        when(cardsStorageImpl.loadCard(MULTIVERSE_ID)).thenReturn(card);
        when(cardsStorageImpl.loadOtherSide(card)).thenReturn(otherSideCard);
        underTest = new CardsInteractorImpl(cardsStorageImpl, logger);
    }

    @Test
    public void testGetLuckyCards() {
        TestObserver<List<MTGCard>> testSubscriber = new TestObserver<>();
        underTest.getLuckyCards(2).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(luckyCards);
        verify(cardsStorageImpl).getLuckyCards(2);
    }

    @Test
    public void testGetFavourites() {
        TestObserver<List<MTGCard>> testSubscriber = new TestObserver<>();
        underTest.getFavourites().subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(favCards);
        verify(cardsStorageImpl).getFavourites();
    }

    @Test
    public void testSaveAsFavourite() {
        MTGCard card = mock(MTGCard.class);
        int[] idFavs = new int[]{1, 2, 3};
        when(cardsStorageImpl.saveAsFavourite(card)).thenReturn(idFavs);
        TestObserver<int[]> testSubscriber = new TestObserver<>();
        underTest.saveAsFavourite(card).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(idFavs);
        verify(cardsStorageImpl).saveAsFavourite(card);
    }

    @Test
    public void testRemoveFromFavourite() {
        MTGCard card = mock(MTGCard.class);
        int[] idFavs = new int[]{3, 4, 5};
        when(cardsStorageImpl.removeFromFavourite(card)).thenReturn(idFavs);
        TestObserver<int[]> testSubscriber = new TestObserver<>();
        underTest.removeFromFavourite(card).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(idFavs);
        verify(cardsStorageImpl).removeFromFavourite(card);
    }

   /* @Test
    public void testLoadSet() {
        TestObserver<List<MTGCard>> testSubscriber = new TestObserver<>();
        underTest.loadSet(set).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(setCards);
        verify(cardsStorageImpl).load(set);
    }*/

    @Test
    public void testLoadIdFav() {
        int[] idFavs = new int[]{6, 7, 8};
        when(cardsStorageImpl.loadIdFav()).thenReturn(idFavs);
        TestObserver<int[]> testSubscriber = new TestObserver<>();
        underTest.loadIdFav().subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(idFavs);
        verify(cardsStorageImpl).loadIdFav();
    }

    @Test
    public void testLoadDeck() {
        TestObserver<List<MTGCard>> testSubscriber = new TestObserver<>();
        underTest.loadDeck(deck).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(deckCards);
        verify(cardsStorageImpl).loadDeck(deck);
    }

    @Test
    public void testDoSearch() {
        TestObserver<List<MTGCard>> testSubscriber = new TestObserver<>();
        underTest.doSearch(searchParams).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(searchCards);
        verify(cardsStorageImpl).doSearch(searchParams);
    }

    @Test
    public void testLoadCardWithMultiverseId() {
        TestObserver<MTGCard> testSubscriber = new TestObserver<>();
        underTest.loadCard(MULTIVERSE_ID).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(card);
        verify(cardsStorageImpl).loadCard(MULTIVERSE_ID);
    }

    @Test
    public void testLoadOtherSideCard() {
        TestObserver<MTGCard> testSubscriber = new TestObserver<>();
        underTest.loadOtherSideCard(card).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(otherSideCard);
        verify(cardsStorageImpl).loadOtherSide(card);
    }
}

File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/interactors/CardsInteractorImplTest.kt
package com.dbottillo.mtgsearchfree.interactors

import com.dbottillo.mtgsearchfree.model.CardsCollection
import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.model.MTGSet
import com.dbottillo.mtgsearchfree.model.SearchParams
import com.dbottillo.mtgsearchfree.model.storage.CardsStorage
import com.dbottillo.mtgsearchfree.util.Logger
import io.reactivex.observers.TestObserver
import io.reactivex.schedulers.Schedulers
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnit
import java.util.*

class CardsInteractorImplTest {

    private val MULTIVERSE_ID = 180607

    @Rule @JvmField
    var mockitoRule = MockitoJUnit.rule()

    lateinit var underTest: CardsInteractor

    @Mock
    lateinit var cardsStorage: CardsStorage
    @Mock
    lateinit var set: MTGSet
    @Mock
    lateinit var searchParams: SearchParams
    @Mock
    lateinit var card: MTGCard
    @Mock
    lateinit var otherSideCard: MTGCard
    @Mock
    lateinit var logger: Logger

    private val favCards = Arrays.asList(MTGCard(3), MTGCard(4))

    @Mock
    lateinit var lukcyCardsCollection: CardsCollection

    @Mock
    lateinit var setCollection: CardsCollection

    @Mock
    lateinit var searchCardsCollection: CardsCollection

    @Mock
    lateinit var schedulerProvider: SchedulerProvider

    @Before
    fun setup() {
        `when`(schedulerProvider.io()).thenReturn(Schedulers.trampoline())
        `when`(schedulerProvider.ui()).thenReturn(Schedulers.trampoline())
        underTest = CardsInteractorImpl(cardsStorage, schedulerProvider, logger)
    }

    @Test
    fun `get lucky cards should call storage and return observable`() {
        `when`(cardsStorage.getLuckyCards(2)).thenReturn(lukcyCardsCollection)
        val testSubscriber = TestObserver<CardsCollection>()
        
        underTest.getLuckyCards(2).subscribe(testSubscriber)
        
        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(lukcyCardsCollection)
        verify(cardsStorage).getLuckyCards(2)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `get favourites should call storage and return observable`() {
        `when`(cardsStorage.getFavourites()).thenReturn(favCards)
        val testSubscriber = TestObserver<List<MTGCard>>()

        underTest.getFavourites().subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(favCards)
        verify(cardsStorage).getFavourites()
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `save card as favourites should call storage and return observable`() {
        underTest.saveAsFavourite(card)

        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verify(cardsStorage).saveAsFavourite(card)
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `remove card as favourites should call storage and return observable`() {
        underTest.removeFromFavourite(card)

        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verify(cardsStorage).removeFromFavourite(card)
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `load set should call storage and return observable`() {
        `when`(cardsStorage.load(set)).thenReturn(setCollection)
        val testSubscriber = TestObserver<CardsCollection>()

        underTest.loadSet(set).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(setCollection)
        verify(cardsStorage).load(set)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `load ifs of favourites should call storage and return observable`() {
        val idFavs = intArrayOf(6, 7, 8)
        `when`(cardsStorage.loadIdFav()).thenReturn(idFavs)
        val testSubscriber = TestObserver<IntArray>()

        underTest.loadIdFav().subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(idFavs)
        verify(cardsStorage).loadIdFav()
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `do search should call storage and return observable`() {
        `when`(cardsStorage.doSearch(searchParams)).thenReturn(searchCardsCollection)
        val testSubscriber = TestObserver<CardsCollection>()

        underTest.doSearch(searchParams).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(searchCardsCollection)
        verify(cardsStorage).doSearch(searchParams)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `load card with multiverse id should call storage and return observable`() {
        `when`(cardsStorage.loadCard(MULTIVERSE_ID)).thenReturn(card)
        val testSubscriber = TestObserver<MTGCard>()

        underTest.loadCard(MULTIVERSE_ID).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(card)
        verify(cardsStorage).loadCard(MULTIVERSE_ID)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

    @Test
    fun `load other side of card should call storage and return observable`() {
        `when`(cardsStorage.loadOtherSide(card)).thenReturn(otherSideCard)
        val testSubscriber = TestObserver<MTGCard>()

        underTest.loadOtherSideCard(card).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(otherSideCard)
        verify(cardsStorage).loadOtherSide(card)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(cardsStorage, schedulerProvider)
    }

}